libresin/libresin/core/sdf_tree/sdf_tree_node.hpp

```diff
@@ -0,0 +1,84 @@
+#ifndef RESIN_SDF_TREE_NODE_HPP
+#define RESIN_SDF_TREE_NODE_HPP
+
+#include <functional>
+#include <libresin/core/id_registry.hpp>
+#include <libresin/core/sdf_shader_consts.hpp>
+#include <libresin/core/sdf_tree/sdf_tree_node_visitor.hpp>
+#include <libresin/core/transform.hpp>
+#include <string>
+#include <unordered_set>
+
+namespace resin {
+
+using SDFBinaryOperation = sdf_shader_consts::SDFShaderBinOp;
+
+struct SDFTreeRegistry;
+class SDFTree;
+
+class SDFTreeNode;
+using SDFTreeNodeId = Id<SDFTreeNode>;
+
+class SDFTreeNode {
+ public:
+  SDFTreeNode() = delete;
+
+  // It is the programmer's responsibility to assert that SDFTreeNode class will not outlive the provided registry!
+  explicit SDFTreeNode(SDFTreeRegistry& tree);
+
+  virtual ~SDFTreeNode();
+
+  virtual std::string gen_shader_code() const               = 0;
+  virtual void accept_visitor(ISDFTreeNodeVisitor& visitor) = 0;
+  virtual std::string_view name() const                     = 0;
+  virtual void rename(std::string&&)                        = 0;
+  [[nodiscard]] virtual std::unique_ptr<SDFTreeNode> copy() = 0;
+  virtual bool is_leaf()                                    = 0;
+
+  inline IdView<SDFTreeNodeId> node_id() const { return node_id_; }
+  inline IdView<TransformId> transform_component_id() const { return transform_id_; }
+  inline Transform& transform() { return transform_; }
+
+  // TODO(migoox): add material component
+
+  inline SDFBinaryOperation bin_op() const { return bin_op_; }
+  inline void set_bin_op(SDFBinaryOperation bin_op) { bin_op_ = bin_op; }
+
+  inline bool has_parent() const { return parent_.has_value(); }
+  inline GroupNode& parent() { return parent_.value(); }
+  inline const GroupNode& parent() const { return parent_.value(); }
+
+  void mark_dirty();
+
+ protected:
+  friend SDFTree;
+  friend GroupNode;
+
+  inline void set_parent(GroupNode& parent) { parent_ = parent; }
+  inline void remove_from_parent() { parent_.reset(); }
+
+  virtual void insert_leaves_to(
+      std::unordered_set<IdView<SDFTreeNodeId>, IdViewHash<SDFTreeNodeId>, std::equal_to<>>& leaves) = 0;
+
+  virtual void remove_leaves_from(
+      std::unordered_set<IdView<SDFTreeNodeId>, IdViewHash<SDFTreeNodeId>, std::equal_to<>>& leaves) = 0;
+
+  virtual void push_dirty_primitives() = 0;
+
+ protected:
+  SDFTreeNodeId node_id_;
+  TransformId transform_id_;
+  Transform transform_;
+  SDFBinaryOperation bin_op_;
+
+  std::optional<std::reference_wrapper<GroupNode>> parent_;
+
+  std::reference_wrapper<SDFTreeRegistry> tree_registry_;
```

Comment:
This comment actually made me think whether we want the nodes to have copy assignment operator and copy constructors explicitly deleted to avoid the node shallow copy. If not then the `std::reference_wrapper` provides the copy assignment out of the box, otherwise we should use the regular reference. 

In my opinion we could actually delete the shallow copies, as I don't see any reason why someone would do the shallow copy of the node instead of saving the id view for the later usage. Moreover in the case of group nodes the shallow copy might be quite expensive, as it involves the collection copies. 

What do you think? @kubazuch @adamieck

Reply:
And the final argument is that the group nodes are already uncopyable as they are the owners of `std::unique_ptr`s, so I think there is no more space for the discussion😊